aptos-move/framework/aptos-experimental/sources/trading/market/pre_cancellation_tracker.move

```diff
@@ -0,0 +1,227 @@
+            expiration_with_order_ids: big_ordered_map::new_with_reusable(),
+            account_order_ids: big_ordered_map::new_with_reusable()
+        }
+    }
+
+    public fun pre_cancel_order_for_tracker(
+        tracker: &mut PreCancellationTracker,
+        account: &signer,
+        client_order_id: u64
+    ) {
+        garbage_collect(tracker);
+        let account_order_id = new_account_client_order_id(signer::address_of(account), client_order_id);
+        if (tracker.account_order_ids.contains(&account_order_id)) {
+            // If the account_order_id already exists with a previously set expiration time,
+            // we update the expiration time.
+            let expiration_time = tracker.account_order_ids.remove(&account_order_id);
+            let order_id_with_expiration =
+                ExpirationAndOrderId { expiration_time, account_order_id };
+            // If the mapping exists, then we remove the order ID with its expiration time.
+            tracker.expiration_with_order_ids.remove(&order_id_with_expiration);
+        };
+        let current_time = aptos_std::timestamp::now_microseconds();
+        let expiration_time = current_time + tracker.pre_cancellation_window_secs;
+        let order_id_with_expiration = ExpirationAndOrderId {
+            expiration_time,
+            account_order_id
+        };
+        tracker.account_order_ids.add(account_order_id, expiration_time);
+        tracker.expiration_with_order_ids.add(order_id_with_expiration, true);
+    }
+
+    public fun is_pre_cancelled(
+        tracker: &mut PreCancellationTracker,
+        account: address,
+        client_order_id: u64
+    ): bool {
+        garbage_collect(tracker);
+        let account_order_id = new_account_client_order_id(account, client_order_id);
+        let expiration_time_option = tracker.account_order_ids.get(&account_order_id);
+        if (expiration_time_option.is_some()) {
+            let current_time = aptos_std::timestamp::now_seconds();
+            let expiration_time = expiration_time_option.destroy_some();
+            if (current_time > expiration_time) {
+                // This is possible as garbage collection may not be able to garbage collect all expired orders
+                // in a single call.
+                tracker.account_order_ids.remove(&account_order_id);
+                let order_id_with_expiration =
+                    ExpirationAndOrderId { expiration_time, account_order_id };
+                tracker.expiration_with_order_ids.remove(&order_id_with_expiration);
+            } else {
+                return true; // Order ID already exists with a valid expiration time.
+            }
+        };
+        return false
+    }
+
+    public fun garbage_collect(tracker: &mut PreCancellationTracker) {
+        let i = 0;
+        let current_time = aptos_std::timestamp::now_seconds();
+        while (i < MAX_ORDERS_GARBAGE_COLLECTED_PER_CALL
```

Comment:
What would happen if the system can't keep up with garbage collect?

Reply:
Every order collects more than 1 of one,  so it should always eventually keep up?

Reply:
Looking at the code, when we can't keep up, there won't be any semantics issues, just taking more storage temporarily right?